excel/src/main/java/org/apache/metamodel/excel/DefaultSpreadsheetReaderDelegate.java

```diff
@@ -148,8 +151,16 @@ private MutableTable createTable(final Workbook wb, final Sheet sheet) {
 
                 for (int j = offset; j < row.getLastCellNum(); j++) {
                     final ColumnNamingContext namingContext = new ColumnNamingContextImpl(table, null, j);
-                    final Column column = new MutableColumn(columnNamingSession.getNextColumnName(namingContext),
-                            ColumnType.STRING, table, j, true);
+                    final Column column;
+                    if (_configuration.isDetectColumnTypes()) {
+                        column =
+                                new MutableColumn(columnNamingSession.getNextColumnName(namingContext), columnTypes[j],
+                                        table, j, true);
+                    } else {
+                        column =
+                                new MutableColumn(columnNamingSession.getNextColumnName(namingContext),
+                                        ColumnType.STRING, table, j, true);
+                    }
                     table.addColumn(column);
                 }
             }
```

Comment:
> I would do this check inside the `getColumnTypes` method, because you only want to execute the logic which scans a number of rows in the Excel sheet for data types when this is activated, otherwise you'll only lose performance for something which you don't use. In case `isDetectColumnTypes` returns false, just have the `getColumnTypes` method return an array filled with `ColumnType.STRING` objects.

you mean this whole part? (lines 146 - 166)